### PR TITLE
Minor fixes to incremental normalization and nesting

### LINKED ISSUE #6161: Normalization: set primary key field as primary key in final table
## Tell us about the problem you're trying to solve
From [slack convo](https://airbytehq.slack.com/archives/C01MFR03D5W/p1631742330376900).
Some users want to use the final table with other tools that require tables have primary key.
> I have additional tables I'm creating in the same db that link to these tables via foreign keys.
I was trying to set up a couple other tools to use as an internal admin portal (directus, nocodb, budibase) and all of them depend on primary keys to different degrees

## Describe the solution you’d like
For simple cases (primary key field is not nested) set the final tables with primary key.

## Describe the alternative you’ve considered or used
A clear and concise description of any alternative solutions or features you've considered or are using today.

## Are you willing to submit a PR?

Remove this with your answer :-)

### PR DESCRIPTION
**Pull request by ChristopheDuong**
## What / How
This PR fixes minor behaviors with incremental:
- choice on when Incremental dbt should be used should depend on destination sync mode, not source sync mode
  - this change enables incremental dbt for FULL_REFRESH-APPEND sync modes too
- safety cast to TIMESTAMP if the schema of raw & normalized tables were changed on `_airbyte_emitted_at` (BigQuery) see https://airbytehq.slack.com/archives/C01MFR03D5W/p1636035876106000
- potential speed improvement on incremental with Postgres for SCD tables (avoid 'view' materialization because it doesn't combine well with dbt incremental on postgres)
- fix some incremental issue with nested tables with array fields (the _airbyte_ab_id column is not unique in that case)
- deduplicated tables in append_dedup sync mode fail to retain only one row per primary key if the cursor field is always NULL (rare edge case)
- unique_key on postgres is now unique for de-duplicated tables, closes https://github.com/airbytehq/airbyte/issues/6161
- moving/re-enabling some tests

## Recommended reading order
1. `x.java`
2. `y.python`

### CODE REVIEW

File: airbyte-integrations/bases/base-normalization/integration_tests/normalization_test_output/postgres/test_nested_streams/models/generated/airbyte_incremental/test_normalization/nested_stream_with_c___long_names_partition.sql [hunk 1]
```diff
@@ -0,0 +1,20 @@
+{{ config(
+    indexes = [{'columns':['_airbyte_emitted_at'],'type':'hash'}],
+    unique_key = '_airbyte_ab_id',
+    schema = "test_normalization",
+    tags = [ "nested" ]
+) }}
```
Comment (ChristopheDuong) [inline airbyte-integrations/bases/base-normalization/integration_tests/normalization_test_output/postgres/test_nested_streams/models/generated/airbyte_incremental/test_normalization/nested_stream_with_c___long_names_partition.sql:6]: Un-nesting an object type column from the parent stream can re-use the same `_airbyte_ab_id` column as unique_key

File: airbyte-integrations/bases/base-normalization/integration_tests/normalization_test_output/postgres/test_nested_streams/models/generated/airbyte_incremental/test_normalization/nested_stream_with_c___names_partition_data.sql [hunk 2]
```diff
@@ -0,0 +1,18 @@
+{{ config(
+    indexes = [{'columns':['_airbyte_emitted_at'],'type':'hash'}],
+    schema = "test_normalization",
+    tags = [ "nested" ]
+) }}
```
Comment (ChristopheDuong) [inline airbyte-integrations/bases/base-normalization/integration_tests/normalization_test_output/postgres/test_nested_streams/models/generated/airbyte_incremental/test_normalization/nested_stream_with_c___names_partition_data.sql:5]: Un-nesting an array type column from the parent stream can NOT re-use the same `_airbyte_ab_id` column as unique_key... each element of the nested array would have the same unique_key and it wouldn't be unique anymore...

File: airbyte-integrations/bases/base-normalization/integration_tests/normalization_test_output/postgres/test_simple_streams/first_output/airbyte_incremental/scd/test_normalization/dedup_exchange_rate_scd.sql [hunk 3]
```diff
@@ -42,15 +42,15 @@ scd_data as (
...
-      case when lag("date") over (
+      case when row_number() over (
         partition by "id", currency, cast(nzd as 
     varchar
 )
         order by
             "date" is null asc,
             "date" desc,
             _airbyte_emitted_at desc
-      ) is null  then 1 else 0 end as _airbyte_active_row,
```
Comment (ChristopheDuong) [inline airbyte-integrations/bases/base-normalization/integration_tests/normalization_test_output/postgres/test_simple_streams/first_output/airbyte_incremental/scd/test_normalization/dedup_exchange_rate_scd.sql:53]: when all (multiple) rows for the same primary key have a cursor (date here) equal to NULL, then this will flag multiple rows as active...

File: airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py [hunk 4]
```diff
@@ -138,15 +138,19 @@ def create_from_parent(
...
+            parent_sync_mode = parent.destination_sync_mode
         result = StreamProcessor.create(
             stream_name=child_name,
             destination_type=parent.destination_type,
             raw_schema=parent.raw_schema,
             default_schema=parent.default_schema,
             schema=parent.schema,
-            # Nested Streams don't inherit parents sync modes?
-            source_sync_mode=SyncMode.full_refresh,
-            destination_sync_mode=DestinationSyncMode.append,
```
Comment (ChristopheDuong) [inline airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py:149]: Sync modes should be propagated to children tables, otherwise incremental is not going to work properly on nested streams

Comment (ChristopheDuong) [inline airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py:149]: Sync modes should be propagated to children tables, otherwise incremental is not going to work properly on nested streams

File: airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py [hunk 5]
```diff
@@ -237,42 +241,55 @@ def process(self) -> List["StreamProcessor"]:
...
                 self.generate_final_model(from_table, column_names),
                 self.get_model_materialization_mode(is_intermediate=False, column_count=column_count),
+                is_intermediate=False,
             )
         else:
+            if DestinationType.POSTGRES.value == self.destination_type:
+                # because of https://github.com/dbt-labs/docs.getdbt.com/issues/335, we have to use tables for postgres
+                forced_materialization_type = TableMaterializationType.TABLE
+            else:
+                forced_materialization_type = TableMaterializationType.VIEW
```
Comment (ChristopheDuong) [inline airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py:270]: VIEW materialization on Postgres and INCREMENTAL don't do well together

**Review by tuliren — APPROVED** [review]
Wow, this is more than a small fix. It fixes so many issue.

It does not seem that all the fixes are unit-tested though (e.g. the case of nested tables with array fields, the rare edge case of deduplicated tables in append_dedup sync mode with NULL cursor field). Can you add test for them?

**Comment by ChristopheDuong** [comment]
> Wow, this is more than a small fix. It fixes so many issue.
> 
> It does not seem that all the fixes are unit-tested though (e.g. the case of nested tables with array fields, the rare edge case of deduplicated tables in append_dedup sync mode with NULL cursor field). Can you add test for them?

yes, It's multiple small or minor fixes...

Most are not unit-testable though (because they rely on data use cases when run with dbt) and unit tests only can test the model generation part.

But they are in the integration tests. I just added the scenario of append_dedup sync mode with the NULL cursor field.

**Comment by tuliren** [comment]
> But they are in the integration tests, some of the fixes surfaced thanks to the tests actually. I just added the scenario of append_dedup sync mode with the NULL cursor field.

Nice. Thanks.

File: airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py (hunk 5 above)
Comment (ChristopheDuong) [reply airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py]: I created this issue to gain a measurable impact of such changes in the future: https://github.com/airbytehq/airbyte/issues/7741
